lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+// 'self.visit(<expr>)' recursion pattern between optional and non-optional inputs.
+extension ASTGenVisitor {
+  @inline(__always)
+  func visit(_ node: TypeSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: ExprSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: GenericParameterClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: (some SyntaxChildChoices)?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: GenericWhereClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: EnumCaseParameterClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
+  }
+
+  @inline(__always)
+  func visit(_ node: InheritedTypeListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node) 
+  }
+
+  @inline(__always)
+  func visit(_ node: PrecedenceGroupNameListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node)
+  }
+}
+
+extension SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      var currentAddress = baseAddress
+      for element in self {
+        currentAddress.initialize(to: transform(element))
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+
+  /// Like ``Sequence.compactMap(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func compactMap<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T?) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      var currentAddress = baseAddress
+      for element in self {
+        guard let transformed = transform(element) else {
+          continue
+        }
+
+        currentAddress.initialize(to: transformed)
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+}
+
+extension Optional where Wrapped: SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  @inline(__always)
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Wrapped.Element) -> T) -> BridgedArrayRef {
+    switch self {
+    case .none: return .init()
+    case .some(let unwrapped): return unwrapped.map(in: astgen, transform)
+    }
+  }
+}
```

Comment:
> I’m not a huge fan of this method because there’s no corresponding stdlib equivalent

What about the methods under `extension Optional<TokenSyntax>`? They’re essentially the same convenience. I think what sets these methods apart is that their result types are arguably `ExpressibleByNilLiteral` (`BridgedArrayRef` is currently not but could follow the precedent set by `BridgedSourceLoc` and `BridgedIdentifier`). Supposing they are, `?? nil` ideally calls for democratization of optional chaining such that it works (flattening and all) with custom `ExpressibleByNilLiteral` types. Another way to make this work for our purposes is to define a custom `map`/`flatMap` on `Optional`, hence the method.

Reply:
I stumbled a bit as well when I read `extension `Optional<TokenSyntax>` but then accepted them because `BridgedIdentifier` is `ExpressibleByNilLiteral` and `?? nil` looks weird. 

I see your argument that you could view `BridgedArrayRef` as `ExpressibleByNilLiteral` now. I’m fine with keeping it.